pytest/node_plugin.py

```diff
@@ -0,0 +1,293 @@
+    Resource Group:	'{name}-rg'
+    Region:		'{location}'
+    Image:		'{vm_image}'
+    Size:		'{vm_size}'"""
+    )
+
+    local.run(f"az group create -n {name}-rg --location {location}")
+    # TODO: Accept EULA terms when necessary. Like:
+    #
+    # local.run(f"az vm image terms accept --urn {vm_image}")
+    #
+    # However, this command fails unless the terms exist and have yet
+    # to be accepted.
+
+    vm_command = [
+        "az vm create",
+        f"-g {name}-rg",
+        f"-n {name}",
+        f"--image {vm_image}",
+        f"--size {vm_size}",
+        f"--boot-diagnostics-storage {boot_storage}",
+        "--generate-ssh-keys",
+    ]
+    # TODO: Support setting up to NICs.
+    if networking == "SRIOV":
+        vm_command.append("--accelerated-networking true")
+
+    data: Dict[str, str] = json.loads(local.run(" ".join(vm_command)).stdout)
+    host = data["publicIpAddress"]
+
+    allow_ping(name)
+    # TODO: Enable auto-shutdown 4 hours from deployment.
+
+    return host, data
+
+
+def delete_vm(name: str) -> None:
+    """Delete the entire allocated resource group."""
+    logging.info(f"Deleting resource group '{name}-rg'")
+    local.run(f"az group delete -n {name}-rg --yes --no-wait")
+
+
+class Node(Connection):
+    """Extends 'fabric.Connection' with our own utilities."""
+
+    name: str
+    data: Dict[str, str]
+
+    def local(self, *args: Any, **kwargs: Any) -> Result:
+        """This patches Fabric's 'local()' function to ignore SSH environment."""
+        return super(Connection, self).run(replace_env=False, env={}, *args, **kwargs)
+
+    @retry(reraise=True, wait=wait_exponential(), stop=stop_after_attempt(3))
+    def get_boot_diagnostics(self, **kwargs: Any) -> Result:
+        """Gets the serial console logs."""
+        # NOTE: Some images can cause the `az` CLI to crash because
+        # their logs aren’t UTF-8 encoded. I’ve filed a bug:
+        # https://github.com/Azure/azure-cli/issues/15590
+        return self.local(
+            f"az vm boot-diagnostics get-boot-log -n {self.name} -g {self.name}-rg",
+            **kwargs,
+        )
+
+    @retry(reraise=True, wait=wait_exponential(), stop=stop_after_attempt(3))
+    def ping(self, **kwargs: Any) -> Result:
+        flag = "-c 1" if platform.system() == "Linux" else "-n 1"
+        return self.local(f"ping {flag} {self.host}", **kwargs)
+
+    def platform_restart(self) -> Result:
+        """TODO: Should this '--force' and redeploy?"""
+        return self.local(f"az vm restart -n {self.name} -g {self.name}-rg")
+
+    def cat(self, path: str) -> str:
+        """Gets the value of a remote file without a temporary file."""
+        with BytesIO() as buf:
+            self.get(path, buf)
+            return buf.getvalue().decode("utf-8").strip()
+
+
+# TODO: The fixtures need to be fixed up since we now have a pair, one
+# for each scope. They need documentation and de-duplication too.
+@pytest.fixture(scope="function")
```

Comment:
In my mind, this is key part of PoC. Fixture cannot control executing order of test cases. It makes a lot of scenario doesn't work. For example (not full list),
1. Some environments (baremetal, hpc) are expensive, if cache an environment to the whole lifecycle of test run (may be hours), v3 will cost more than v2. If release environment each time, that means test suite will spend more time to deploy more new environments. 
2. there are quota limitation for each subscription. Each subscription has quota limitation on each location, it’s hit many times currently.

Reply:
It doesn't need to? This fixture is just to get a connection to a node (and for now also handles deploying the node). I don't see why any of this couldn't be supported.